Prefer user-provided node IP

stephenfin opened the pull request:
Place primary node IP first in list of node addresses

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->
/kind bug

#### What this PR does / why we need it:

As discussed in kubernetes/cloud-provider#56, CCM no longer prefers the
node IP passed by the user to `kublet` (via the `--node-ip` parameter). This PR
corrects that oversight by placing the annotated `NodeIP` first in the returned
list of `NodeAddresses` stored in `node.Status.Addresses`.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes kubernetes/cloud-provider#56

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

```release-note
A node IP provided to kublet via `--node-ip` will now be preferred for
when determining the node's primary IP and using the external cloud provider
(CCM).
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dims commented:
/assign @andrewsykim

mdbooth commented:
/ok-to-test

mdbooth commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,17 +693,35 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
+			return nodeIP, nodeIPExists
+		}
+
+		if index == 0 {
+			return &nodeAddresses[index], nodeIPExists
 		}
+
+		primary := nodeAddresses[index]
+		copy(nodeAddresses[1:index+1], nodeAddresses[0:index])
+		nodeAddresses[0] = primary
+		nodeIP = &primary
```
It's not very complicated, but I'd still put a comment or two in here.

mdbooth commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,17 +693,35 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
 
+// ensureNodeProvidedIPExists will check if the nodeIP suggested by the user is
+// present and move it to the front of the list of node IP addresses if so
 func ensureNodeProvidedIPExists(node *v1.Node, nodeAddresses []v1.NodeAddress) (*v1.NodeAddress, bool) {
```
Worth renaming this? It's no longer just ensuring.

andrewsykim replied:
This function probably needs to be renamed now too, because we're no longer just ensuring the node IP is there, but we're also shuffling the order.

mdbooth reviewed (changes requested):
I think this change needs to go in kubernetes/cloud-provider directly rather than kubernetes/kubernetes.

stephenfin commented:
> I think this change needs to go in kubernetes/cloud-provider directly rather than kubernetes/kubernetes.

Nah, [`CONTRIBUTING.md` in that repo says this is where it should go](https://github.com/kubernetes/cloud-provider/blob/master/CONTRIBUTING.md), oddly enough

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,17 +693,35 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
+// ensureNodeProvidedIPExists will check if the nodeIP suggested by the user is
+// present and move it to the front of the list of node IP addresses if so
 func ensureNodeProvidedIPExists(node *v1.Node, nodeAddresses []v1.NodeAddress) (*v1.NodeAddress, bool) {
+	find := func(address string) (int, bool) {
```
nit: find -> findIndex

mdbooth commented:
@stephenfin I've just pushed an update addressing review comments as you're out today. Please squash my commit or drop it as you see fit when you're back in.

JoelSpeed commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -483,11 +485,14 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(
+	// in the cloud provider and place it first in the node addresses list
+	// before removing the taint on the node
+	nodeAddresses := preferProvidedNodeIP(node, instanceMeta.NodeAddresses)
+	if nodeAddresses == nil {
```
This is side-effecty and non-obvious, would be better to return an explicit bool or error when we don't find the address wouldn't it? Can `instance.NodeAddresses` ever be nil? If it is, there could be multiple ways that `nodeAdresses` becomes nil which makes this even more like guesswork

mdbooth replied:
NodeAddresses *should* never be nil, but now that you've said it I don't think it's good practise to assume that here and we should handle it.

I agree it's side-effecty, although returning the input parameter was my attempt to make this more obvious. The previous iteration just returned a bool, and I didn't think that was obvious at all that it had a side-effect. Perhaps if it was in the name: `modifyNodeAddressesForPreferredNodeIP() -> bool(ok)`? That would also have the advantage of gracefully handling nil NodeAddresses 🤔

JoelSpeed replied:
> returning the input parameter was my attempt to make this more obvious

This is good, and there's no reason not to return two return variables right? You could return the list and a bool, or a list and an error, I could see either of those making sense here depending how you look at it

JoelSpeed commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:
```diff
@@ -162,6 +162,121 @@ func TestEnsureNodeExistsByProviderID(t *testing.T) {
+					Type:    v1.NodeInternalIP,
+					Address: "192.168.17.8",
+				},
+				{
+					Type:    v1.NodeInternalIP,
+					Address: "192.168.25.1",
+				},
```
This test case only swaps two items, could we add a couple more so that we can see it successfully moving a few addresses around?

mdbooth replied:
Yep. Lets have the major cases where node IP is:

* First
* Middle
* Last

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:
```diff
@@ -162,6 +162,121 @@ func TestEnsureNodeExistsByProviderID(t *testing.T) {
 	}
 }
 
+func Test_ensureNodeProvidedIPExists(t *testing.T) {
```
Test_preferProvidedNodeIP

mdbooth commented:
/hold @stephenfin I've pushed a second commit for the new review comments. Please only remove this hold if you're happy with my changes.

JoelSpeed commented:
Changes look good, thanks for the updates @mdbooth
/lgtm

stephenfin commented:
> /hold @stephenfin I've pushed a second commit for the new review comments. Please only remove this hold if you're happy with my changes.

Looks good to me. I simply squashed them.

/unhold

JoelSpeed commented:
/lgtm

mdbooth commented:
/test pull-kubernetes-e2e-gce-ubuntu-containerd

elmiko reviewed (commented):
we talked about this at the sig cloud provider meeting tonight, i am +1 for the approach proposed here.

one thing we noted is that this _might_ make some breakages for code that was expecting some /other/ ordering. but imo, those breakages will show us places that could use strengthening.

/lgtm

mdbooth commented:
I just verified this on an actual cluster. I can confirm both that the specified node-ip is placed first in the list of Node addresses, and that this allows the node to come up successfully.

```
$ oc -n openshift-machine-api get machine/mbooth-psi-jqzwm-worker-0-n789b -o json | jq .status.addresses
[
  {
    "address": "192.168.200.135",
    "type": "InternalIP"
  },
  {
    "address": "10.0.130.164",
    "type": "InternalIP"
  },
  {
    "address": "mbooth-psi-jqzwm-worker-0-n789b",
    "type": "Hostname"
  },
  {
    "address": "mbooth-psi-jqzwm-worker-0-n789b",
    "type": "InternalDNS"
  }
]
```
```
$ oc get node/mbooth-psi-jqzwm-worker-0-n789b -o json | jq .status.addresses
[
  {
    "address": "10.0.130.164",
    "type": "InternalIP"
  },
  {
    "address": "192.168.200.135",
    "type": "InternalIP"
  },
  {
    "address": "mbooth-psi-jqzwm-worker-0-n789b",
    "type": "Hostname"
  }
]
```

mdbooth commented:
> > one thing we noted is that this _might_ make some breakages for code that was expecting some /other/ ordering. but imo, those breakages will show us places that could use strengthening.
> 
> if those places are outside our control, and outside the control of the user running the cluster, what is the recourse?

This patch has no effect unless the user has specified `--node-ip` to kubelet. Without this patch that argument is effectively ignored when using a CCM. To get the previous behaviour they can remove the `--node-ip` argument to kubelet.

elmiko commented:
> > > one thing we noted is that this _might_ make some breakages for code that was expecting some /other/ ordering. but imo, those breakages will show us places that could use strengthening.
> > 
> > 
> > if those places are outside our control, and outside the control of the user running the cluster, what is the recourse?
> 
> This patch has no effect unless the user has specified `--node-ip` to kubelet. Without this patch that argument is effectively ignored when using a CCM. To get the previous behaviour they can remove the `--node-ip` argument to kubelet.
> 
> I think that in order for this to break the user they would have to be explicitly passing an incorrect `--node-ip` to kubelet. If the user is either not passing `--node-ip`, or has set it correctly, they will not be affected by this change.
> 
> I'm not familiar with this process yet, but that sounds like it might warrant a release note? If so, could you point me in the right direction to document it?

yeah, at some level this is an academic debate. in the sig meeting we were discussing if this patch could potentially break something that is already in the wild. i posited that it is possible that somewhere is a controller that was expecting the ordering, based on who knows what, and that those implementations might be broken by this.

firstly, that's a lot of "if"s and "might"s, but secondly, i believe if people are doing this already then this change would indicate that they were using a fragile method for parsing the ip addresses anyway, and this change would point out those places that need fixing.

as for the recourse, if this code does exist, and it is broken because of this patch, then those place would probably want to start using this patch so they could be more deterministic about the ip addresses they are choosing. also a win in my book.

nckturner commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -483,11 +485,14 @@ func (cnc *CloudNodeController) getNodeModifiersFromCloudProvider(
+	// in the cloud provider and place it first in the node addresses list
+	// before removing the taint on the node
+	nodeAddresses, ok := preferProvidedNodeIP(node, instanceMeta.NodeAddresses)
+	if !ok {
```
Given that these logs might be viewed after the node object has gone away or changed, it would make debugging easier if we also returned the node IP that was not present in the list of node addresses, and printed it in the "failed to find kubelet node IP..." error message.

mdbooth replied:
Addressed in latest patch, thanks.

mdbooth commented:
@nckturner @andrewsykim Any chance this is good to go?

elmiko reviewed (commented):
/lgtm

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,19 +698,43 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
+
+	// Move the suggested address to the front by moving those before it right one position
+	primary := nodeAddresses[index]
+	copy(nodeAddresses[1:index+1], nodeAddresses[0:index])
```
Had a conversion with Jordan today about the implications of this change. Overall it seems correct that node controller sets the order of the address, but it should set the order only scoped to the address type it belongs to. E.g. if --node-ip matches against an InternalIP, it should sort it such that it is the first InternalIP in status but not necessary the first amongst all adddresses (unless it was already the case).

mdbooth replied:
Thanks. I've just pushed a new commit (well 2, but one of them is just readability).

Incidentally, it's only a couple of lines of code but I personally feel future me would appreciate the verbose comment, especially if I hadn't had coffee yet. Hope it doesn't fall foul of style guidelines.

danwinship replied:
Wouldn't it make more sense to extract the kubelet cloud-provider-node-ip-addresses-reordering code to `"k8s.io/component-helpers/node/util"` and then use that code here?

mdbooth replied:
I don't imagine there would be any other users of it, or are you thinking it would just be tidier?

danwinship replied:
There wouldn't be any other users. And I guess kubelet's usage of it will eventually go away when the legacy providers go away... So maybe just copy the code from there to here instead?

It's just, if we want the same behavior, let's have the same code. (The initial version of the code here didn't do the right thing... are we sure the current version does?)

mdbooth replied:
Now I understand: have kubelet legacy and cloud-provider use the same re-ordering code, which is in node/util. I'll look into it.

liggitt commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,19 +698,62 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
+		return 0, false
+	}
+
+	nodeIPIndex, nodeIPFound := findNodeIPIndex(providedIP)
```
this is only reordering a single record... the cloud provider could return multiple addresses with Address=<nodeIP> with different types

mdbooth replied:
Oh, that's fun.

mdbooth replied:
Thinking about it, I think we should confine this behaviour to internal and external IP. I don't think it's relevant to the others? However, I can implement re-ordering for both if they both match.

liggitt replied:
I don't really have context to know what the desired/expected behavior is, but scoping re-ordering to a particular type/types could be reasonable...

mdbooth replied:
This can now re-order multiple type matches, but explicitly confines the behaviour to internal and external IP.

mdbooth commented:
/test pull-kubernetes-e2e-kind

stephenfin commented:
/retest-required

stephenfin commented:
@liggitt Anything more needed on this to get it over the line?

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -349,10 +349,12 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
+	// (consistent with the behaviour in kubelet)
+	nodeAddresses, err := preferProvidedNodeIP(node, nodeAddresses)
+	if err != nil {
+		klog.Errorf("error handling provided node IP: %v", err)
```
Keep the node name in this log line

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,19 +698,83 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
+		// No suggested node IP
+		return nodeAddresses, nil
+	}
+	klog.V(2).Infof("Kubelet provided node IP %s. Ensuring it is listed as first node address.", providedIP)
```
nit: `Ensuring it is listed as first node address by type`

mdbooth commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller_test.go:
```diff
@@ -162,6 +162,311 @@ func TestEnsureNodeExistsByProviderID(t *testing.T) {
 	}
 }
 
+func Test_preferProvidedNodeIP(t *testing.T) {
```
Added.

andrewsykim commented:
/cc @thockin @bowei @aojea @danwinship 

(from SIG call today, thought it would be good to get SIG Network review on this PR)

danwinship commented:
Oh, so actually, another thing from [KEP-1664](https://github.com/kubernetes/enhancements/pull/1665): if we don't trust the external cloud provider to know the best primary IP, then we can't trust it to know the best secondary / other-IP-family IP either, so we need to allow the user to pass a dual-stack value to `--node-ip`. eg:

- The cloud says the IPs are `[ 1.2.3.4, 5.6.7.8, fd01::1234, fd01::5678 ]` and the admin does nothing special, so kubelet uses `1.2.3.4` as the primary node IP and `fd01::1234` as the secondary node IP.
- The cloud says the IPs are `[ 1.2.3.4, 5.6.7.8, fd01::1234, fd01::5678 ]` but the admin wants to use `5.6.7.8` and `fd01::5678` as the primary and secondary node IPs, so they'd pass `--node-ip 5.6.7.8,fd01::5678`, which kubelet would then pass to the external cloud provider, which would then reorder the list to `[5.6.7.8, fd01::5678, 1.2.3.4, fd01::1234]` and set `node.status.addresses` to that, and kubelet would then use `5.6.7.8` and `fd01::5678` as the node IPs.
- The cloud says the IPs are `[ 1.2.3.4, 5.6.7.8, fd01::1234, fd01::5678 ]` and the admin wants to use `1.2.3.4` and `fd01::1234`, but they want the cluster to be IPv6-primary, not IPv4-primary, so they'd pass `--node-ip fd01::1234,1.2.3.4` to kubelet, which kubelet would pass to the cloud provider, which would then set `node.status.addresses` to `[fd01::1234, 1.2.3.4, 5.6.7.8, fd01::5678]`, and then kubelet would use `fd01::1234` as the primary and `1.2.3.4` as the secondary.

In the current kubelet code, it accepts dual `--node-ip` only when `--cloud-provider` is unset. So that would need to be changed, and you'd either have to pass both node-ip values to the cloud provider in the existing node-ip annotation, or add a new annotation for the second value. (And the current kubelet node-IP-reordering code only handles a single IP, so you'll have to update it to be able to deal with two IPs.)

(In theory we could support dual `--node-ip` with legacy cloud providers too; the reason I didn't add it [when I added bare-metal dual `--node-ip` support](https://github.com/kubernetes/kubernetes/pull/95239) was because we hadn't decided what to do about the legacy-vs-external `--node-ip` behavior difference yet. OTOH (a) I'm not sure dual `--node-ip` support would be useful with any of the legacy cloud providers given that most of them don't really support dual-stack and (b) I think there's a push toward "make people migrate to external cloud providers if they want new features" anyway?)

mdbooth commented:
@danwinship Thanks, that's really insightful. To be purely pragmatic I'd like to break this down into:

* Things I need to do 'today'
* Ways I might improve it 'tomorrow'
* How I still think we've got an architectural problem to fix here

#### Today:

I like your idea of re-using the legacy code for the cloud-provider. e.g. My code doesn't handle a nodeIP of 0.0.0.0. The precise interface will be mechanical and the details will fall out in the refactoring, but essentially we should pass in nodeIP, secondaryNodeIP, and other parts as necessary, and have it spit out a resulting list. This should be equally applicable to CCM and kubelet.

secondaryNodeIP will always be empty for external CCM because we currently have nowhere to stash it. I'd worry about changing the semantics of the existing annotation to allow multiple addresses because I'd have to think too hard about upgrades, and there's always a corner case I forget when I try to do that.

External CCM will be confined to a single --node-ip. This will solve the problem for most affected people. I'd like to aim to get this merged asap.

#### Maybe tomorrow:

Add a second annotation to allow passing a second --node-ip to the cloud provider, and pass that as secondaryNodeIP to the same refactored code we're already using. This would bring the behaviour of CCM on a par with legacy cloud provider. There's no upgrade issue because old CCM safely ignores an annotation it doesn't know about.

The code for this will be trivial, so I'd be happy to do this up front as long as there's no additional process around adding a new annotation. I'm keep to solve the 'make it work' problem for most people asap, so if this will add delay I'd like to merge the easy bit while we contemplate it, as that will fix the majority of the problem anyway.

#### The architecture problem:

I don't think the cloud provider knows the answer to this question. The question is fundamentally an installation question, not a runtime question, so IMHO the canonical source should be the installer. The 2 I work on are CAPI and OpenShift. Currently neither of those know the answer either, but by comparison it would be relatively simple to add some metadata there. I think we currently get away with it through a combination of:

* Most people deploy fairly simple network architectures
* People deploying more complex network architectures are relying on undefined behaviour

In a CAPI model, I think the most logical place for this information to exist is on the Machine (or ProviderMachine) object. CAPI knows not only which networks were added, but also *why* they were added. The cloud provider knows which networks were added, but doesn't have that additional context about their intended use unless it's inferred from something like ordering. Ordering is clearly not sufficient in all real-world use cases, which is why we have ended up with --node-ip.

I'm definitely not going to land that fix in the short term, though 😂 It's very much on my radar, but today I'm going to concentrate on the simpler described above if you think it's sufficient and palatable.

danwinship commented:
yeah, I guess solving just the single-stack problem does move us in the direction of the complete solution, so it's reasonable to solve just that problem now

mdbooth committed:
cloud-provider handles kubelet's --node-ip

When using a legacy cloud provider, if kubelet is passed a node address
in --node-ip it will use this address in preference out the the
addresses by the cloud provider.

When using an external cloud provider, kubelet will annotate the Node
with the first --node-ip for use by the cloud provider. The cloud
provider validates this annotation but does not otherwise use it,
meaning that --node-ip has no effect.

This change moves the node address filtering code from kubelet to
component-helpers and updates both kubelet and cloud-provider to use it.
There is no functional change to kubelet, but cloud-provider now honours
kubelet's --node-ip.

mdbooth commented:
@danwinship As discussed, this new commit is a complete rewrite to re-use the code from kubelet in cloud-provider.

Interestingly, while moving this code I noticed that we don't actually use secondary IP unless... I think that branch bootstraps Addresses for external cloud provider? So I suspect it's not actually relevant after all. IIUC it's not a regression from legacy cloud provider in any case, which was the major issue here.

I have deliberately left the duplicate unit tests in kubelet's `setters_test.go`, but these should be removed. Assuming they pass CI (they do locally) this should give us confidence that the code motion hasn't introduced a regression. I have copied the relevant unit tests to component-helpers. If we were using gerrit I would have immediately submitted a second commit to remove them and we could be satisfied that CI ran against both commits, but this doesn't work with GitHub's PR model. I'll have to submit a follow-up PR to remove them.

danwinship reviewed (commented):
lgtm... there are some icky bits in the existing node-ip-handling code, but since the goal is to preserve the existing behavior, it makes sense to not bother trying to make the code nicer right now.

danwinship commented:
> Interestingly, while moving this code I noticed that we don't actually use secondary IP unless... I think that branch bootstraps Addresses for external cloud provider? So I suspect it's not actually relevant after all. IIUC it's not a regression from legacy cloud provider in any case, which was the major issue here.

Yeah, I think I mentioned that above; kubelet only even allows you to have a secondaryNodeIP in the bare-metal case, so the cloud case doesn't look at it. So not a regression, just a missing feature that we will need at some point in the future.

mdbooth commented:
@andrewsykim Are you able to help this across the line?

andrewsykim commented on pkg/kubelet/nodestatus/setters.go:
```diff
@@ -115,56 +116,9 @@ func NodeAddress(nodeIPs []net.IP, // typically Kubelet.nodeIPs
-			} else {
-				// If nodeIP is unset, just use the addresses provided by the cloud provider as-is
-				nodeAddresses = cloudNodeAddresses
+			nodeAddresses, err := nodeutil.PreferNodeIP(nodeIP, cloudNodeAddresses)
```
Sorry, I must have missed this from one of the other reviews, but what's the justification for including this behvior change in kubelet as well? I thought it was initially for external providers only when using --node-ip?

andrewsykim replied:
Okay after reading through the context I see it was so we can re-use similar code in kubelet. Does that mean we are introducing a behavior change in kubelet too?

andrewsykim replied:
Okay after reading through the context I see it was so we can re-use similar code in kubelet. Does that mean we are introducing a behavior change in kubelet too?

andrewsykim replied:
sorry -- missed your last message with Github being a bit flaky today. Sounds good

andrewsykim commented:
> @andrewsykim Are you able to help this across the line?

This PR is on my radar and will make sure it doesn't slip v1.24

andrewsykim commented on staging/src/k8s.io/component-helpers/node/util/nodeaddresses.go:
```diff
@@ -0,0 +1,96 @@
+// * If nodeIP matches an address of a particular type (internal or external),
+//   that will be the *only* address of that type returned.
+// * All remaining addresses are listed after.
+func PreferNodeIP(nodeIP net.IP, cloudNodeAddresses []v1.NodeAddress) ([]v1.NodeAddress, error) {
```
> It's now called by both kubelet and cloud-provider, so I don't think we can move it into cloud provider. 

But it's only the part of the kubelet dealing with cloud node addresses right? Or does it also pertain to --node-ip without the cloud provider?

andrewsykim replied:
> It's the code in kubelet which re-orders node addresses when not using an external cloud provider.

When not using an external cloud provider, but still using the internal/in-tree one though right? k8s.io/cloud-provider can still contain helper functions used by in-tree cloud provider code

andrewsykim replied:
If the usage of PreferNodeIP can be used when kubelet does not set the --cloud-provider flag, then k8s.io/component-helpers makes sense to me. If it's exclusively used when cloud provider is enabled (external or internal), then k8s.io/cloud-provider probably makes more sense since the kubelet will (eventually) not use that code at some point

mdbooth replied:
@andrewsykim Sorry, I hadn't understood. Yes, this code will only be called when there is *some* cloud provider, either internal or external.

@danwinship @mrunalp Any reason I should not respin this and move the node address code to the suggested location in cloud provider instead? We already have deps on cloud-provider and cloud-provider/api in this file, so I assume this wouldn't be problematic?

danwinship replied:
I have no opinion on where it belongs...

mdbooth commented:
This latest push is identical to the previous one, but moves the node addresses code from component-helpers to cloudprovider/node/helpers

mdbooth commented:
/test pull-kubernetes-integration

andrewsykim commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -693,19 +712,19 @@ func nodeAddressesChangeDetected(addressSet1, addressSet2 []v1.NodeAddress) bool
+func getNodeProvidedIP(node *v1.Node) (net.IP, error) {
+	providedIP, ok := node.ObjectMeta.Annotations[cloudproviderapi.AnnotationAlphaProvidedIPAddr]
+	if !ok {
+		// No provideded node IP
```
^ this still needs to be fixed

andrewsykim replied:
(just removing the comment would be fine as well TBH)

andrewsykim reviewed (commented):
/approve

Just some minor nits left

esiperu-est commented:
Hi,
will this change be available in 1.24?

mdbooth commented:
> Hi, will this change be available in 1.24?

As I made a couple of changes to address nits I'm going to wait another 24 hours before hunting for an lgtm on this, but it's currently looking promising!

mdbooth commented on staging/src/k8s.io/cloud-provider/controllers/node/node_controller.go:
```diff
@@ -349,12 +351,20 @@ func (cnc *CloudNodeController) updateNodeAddress(ctx context.Context, node *v1.
+	// If kubelet provided a node IP, prefer it in the node address list
+	nodeIP, err := getNodeProvidedIP(node)
+	if err != nil {
+		klog.Errorf("Failed to get preferred node IP for node %q: %w", node.Name, err)
```
Pretty sure it doesn't, but that's how my fingers are wired. Change incoming.

mdbooth replied:
Updated them both.

mdbooth commented:
/test pull-kubernetes-integration

andrewsykim reviewed (commented):
/lgtm

Thanks for iterating on this @mdbooth

mdbooth commented:
/hold cancel

mdbooth commented:
/retest-required

mdbooth commented:
/retest-required

mdbooth commented:
From the job history, looks like pull-kubernetes-verify and pull-kubernetes-integration are failing for unrelated reasons. Going to leave retesting for a bit.

mdbooth commented:
Looks like this was https://github.com/kubernetes/kubernetes/pull/108901, which has now merged

/retest-required

mdbooth commented:
/retest-required

ehashman commented:
/priority important-soon
/triage accepted

mdbooth commented:
/retest
